File: tasks/2026/round1/honestpaths.cpp
// @check-accepted: examples Nsmall GraphSmall Tree UnitWeights no-limits
#include <vector>
#include <array>
#include <queue>
#include <stdio.h>
#include <iostream>
#include <algorithm>
#include <utility>
#include <numeric>
using namespace std;
typedef long long ll;
#define pb push_back
#define all(x) x.begin(), x.end()
typedef array<ll, 2> ll2;
int main(){
  ios_base::sync_with_stdio(false);
  cin.tie(0);
  cout.tie(0);

  int n, m, trg;
  cin >> n >> m;
  trg = n;
  vector<vector<ll2>> vec(n + 1);
  vector<ll> dist(n + 1, (ll)1e15);
  dist[trg] = 0;
  vector<ll> dp(n + 1);
  vector<ll> nxt(n + 1);

  for (int i = 1; i <= m; i++){
    ll x, y, w;
    cin >> x >> y >> w;
    vec[x].pb({w, y});
    vec[y].pb({w, x});
  }
  priority_queue<ll2, vector<ll2>, greater<ll2>> pq;
  pq.push({0, trg});
  while (!pq.empty()){
    auto [cur, nod] = pq.top();
    pq.pop();
    if (dist[nod] != cur) continue;
    for (auto [w, y] : vec[nod]){
      if (w + cur < dist[y]){
        dist[y] = cur + w;
        pq.push({dist[y], y});
      }
    }
  }

  vector<int> ord(n);
  dp[trg] = 0;
  iota(all(ord), 1);
  sort(all(ord), [&] (ll a, ll b){
    return dist[a] < dist[b];
  });
  ll rez = trg;
  for (auto x : ord){
    for (auto [w, y] : vec[x]){
      if (dist[y] >= dist[x]) continue;
      if (dp[y] + w > dp[x]){
        dp[x] = dp[y] + w;
        nxt[x] = y;
      }
    }
    if (dp[x] > dp[rez]){
      rez = x;
    }
  }
  vector<int> ans;
  int x = rez;
  do{
    ans.pb(x);
    x = nxt[x];
  } while (x != 0);
  cout << dp[rez] << ' ';
  cout << ans.size() << '\n';
  for (auto x : ans) cout << x << ' ';
  cout << '\n';
  return 0;
}

File: tasks/2026/round1/honestpaths.py
#!/usr/bin/env python3
# @check-accepted: examples Nsmall GraphSmall
# @check-time-limit-exceeded: Tree UnitWeights no-limits

import sys
import heapq

data = list(map(int, sys.stdin.buffer.read().split()))
it = iter(data)
n = next(it)
m = next(it)
trg = n
vec = [[] for _ in range(n + 1)]
INF = 10**18
dist = [INF] * (n + 1)
dist[trg] = 0
dp = [0] * (n + 1)
nxt = [0] * (n + 1)
for _ in range(m):
    x = next(it)
    y = next(it)
    w = next(it)
    vec[x].append((w, y))
    vec[y].append((w, x))
pq = [(0, trg)]
while pq:
    cur, nod = heapq.heappop(pq)
    if dist[nod] != cur:
        continue
    for w, y in vec[nod]:
        if w + cur < dist[y]:
            dist[y] = w + cur
            heapq.heappush(pq, (dist[y], y))
ord_nodes = list(range(1, n + 1))
ord_nodes.sort(key=lambda a: dist[a])
rez = trg
for x in ord_nodes:
    for w, y in vec[x]:
        if dist[y] >= dist[x]:
            continue
        val = dp[y] + w
        if val > dp[x]:
            dp[x] = val
            nxt[x] = y
    if dp[x] > dp[rez]:
        rez = x
ans = []
x = rez
while x != 0:
    ans.append(x)
    x = nxt[x]
out = []
out.append(f"{dp[rez]} {len(ans)}")
out.append(" ".join(map(str, ans)))
sys.stdout.write("\n".join(out) + "\n")
